packages/core/src/common/test/mock-logger.ts

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (C) 2017 Ericsson and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { injectable } from 'inversify';
+import { ILogger, Loggable } from '../logger';
+
+@injectable()
+export class MockLogger implements ILogger {
```

Comment:
I thought we would not need such things with sinon?

Reply:
Unfortunately sinon can't create a mock class from the interface seems it's not possible to do that in TypeScript

Reply:
See this post about it: https://stackoverflow.com/questions/37027776/how-to-stub-a-typescript-interface-type-definition

Note we could use something like:
```typescript
testContainer.bind(ILogger).toDynamicValue(ctx => {
        const logger = <ILogger><any>{ warn: function {} };
        sinon.stub(logger, 'warn').callsFake(() => { });
        return logger;
    });
```

But then we would have to implement something each time we add a call to a logger function in the code and the sinon.stub code would be hard to reuse.

Reply:
I think we do that all the time with the JsonRPC proxies.

Reply:
Ah I see, sinon cannot reflect on the declarations, of course. In that case it is pretty useless IMHO. I like the previous approach with an abstract class and one template method much better.

Reply:
There's a few problems with the abstract class approach:

- If you generalize this it means every dependency of a tested class in theia needs an abstract version without any dependencies.

- The abstract class is not needed by the business logic in theia aside for the tests and thus is adding complexity in the wrong domain, thus contributing to bloat.  The complexity should be added where it is needed, namely in the tests. 

- The abstract class may be non-trivial to write since it may not be easy to write a useful abstract class without the dependencies.

I mean in this case the logger abstract class is meant to be really working without its dependencies, and could be used if there was a need for it in the business logic of theia.

But doing such a class for everything that needs a mock will be non-trivial and may lead to complex dead code if it's only used to provide mocks for testing.

It's much easier to just write a mock that implements the class.

- Also the mocked derived class from this abstract class can't be adapted for multiple scenarios.  Consider if you a class like:

```typescript
class Tested {
  constructor (protected dep: Dependency) {
  }

  test() {
    try {
      this.dep.foo();
    } catch (e) {
      console.log(e);
    }
  }
}
```

Now you want to make a mock for this.dep.foo() such that you can test if it throws an exception and if it doesn't.

Using the abstract class / derived mock approach you will need to create 2 mocks that subclass the foo() method differently.

Sinon is very useful for this use-case, you can stub the foo() method and have it return what you need for each test easily.

This can be expanded also to other requirements like checking the args passed to that function or making sure it is called a number of times etc...

Doing these things without sinon would be a pain.


WDTY?

Reply:
Yes, I am happy to use sinon in such cases, too. But this change adds quite some code (+401 −138) on the one side and looks neither simpler nor more readable to me. Anyway, this concrete change is ok for me both ways. So please go ahead if you like it better. 

What is important to me, is to have the freedom to use what works best on a case by case and not always being forced to use sinon and inversify if it doesn't help.

Reply:
It adds more code because of the Mock but this will code will be reused. For example by the keybindings soon.

It's not ideal for me that we could not use DI but I'm willing to let it go.

However can we agree that we should not abstract classes for test purposes only? And rather use mocks like I did ?

Reply:
Isn't that one of the main use cases of DI ? to be able to switch the implementation for mocks?

Reply:
That is one advantage, yes, which boils down to delegation instead of inheritance.